edward/inferences/sghmc.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import six
+import tensorflow as tf
+
+from edward.inferences.monte_carlo import MonteCarlo
+from edward.models import Normal, RandomVariable, Empirical
+from edward.util import copy
+
+
+class SGHMC(MonteCarlo):
+  """Stochastic gradient Hamiltonian Monte Carlo (Chen et al., 2014).
+
+  Notes
+  -----
+  In conditional inference, we infer :math:`z` in :math:`p(z, \\beta
+  \mid x)` while fixing inference over :math:`\\beta` using another
+  distribution :math:`q(\\beta)`.
+  ``SGHMC`` substitutes the model's log marginal density
+
+  .. math::
+
+    \log p(x, z) = \log \mathbb{E}_{q(\\beta)} [ p(x, z, \\beta) ]
+                \\approx \log p(x, z, \\beta^*)
+
+  leveraging a single Monte Carlo sample, where :math:`\\beta^* \sim
+  q(\\beta)`. This is unbiased (and therefore asymptotically exact as a
+  pseudo-marginal method) if :math:`q(\\beta) = p(\\beta \mid x)`.
+  """
+  def __init__(self, *args, **kwargs):
+    """
+    Examples
+    --------
+    >>> z = Normal(mu=0.0, sigma=1.0)
+    >>> x = Normal(mu=tf.ones(10) * z, sigma=1.0)
+    >>>
+    >>> qz = Empirical(tf.Variable(tf.zeros([500])))
+    >>> data = {x: np.array([0.0] * 10, dtype=np.float32)}
+    >>> inference = ed.SGHMC({z: qz}, data)
+    """
+    super(SGHMC, self).__init__(*args, **kwargs)
+
+  def initialize(self, step_size=0.25, friction=0.1, *args, **kwargs):
+    """
+    Parameters
+    ----------
+    step_size : float, optional
+      Constant scale factor of learning rate.
+    friction : float, optional
+      Constant scale on the friction term in the Hamiltonian system.
+    """
+    self.step_size = step_size
+    self.friction = friction
+    self.v = {z: tf.Variable(tf.zeros(qz.params.get_shape()[1:]))
+              for z, qz in six.iteritems(self.latent_vars)}
+    return super(SGHMC, self).initialize(*args, **kwargs)
+
+  def build_update(self):
+    """
+    Simulate Hamiltonian dynamics with friction using a discretized
+    integrator. Its discretization error goes to zero as the learning rate
+    decreases.
+    Implements the update equations from (15) of Chen et al., 2014.
+    """
+    # TODO: Would it be more memory-efficient to use tf.control_dependencies,
+    # rather than repeatedly re-assign attributes stored on self.v?
+    old_sample = {z: tf.gather(qz.params, tf.maximum(self.t - 1, 0))
+                  for z, qz in six.iteritems(self.latent_vars)}
+    old_v_sample = {z: v for z, v in six.iteritems(self.v)}
+
+    # Simulate Hamiltonian dynamics with friction.
+    friction = tf.constant(self.friction, dtype = tf.float32)
+    # TODO: Allow option for exponentially decaying learning rate, or similar.
```

Comment:
Got it re: tensorflow.

I'll make a new PR that adds an option to keep the learning rate constant, along with an example showing cases where this may be a bit easier to work with.